sdlib/d/gc/arena.d

```diff
@@ -148,6 +148,43 @@ public:
 		return null;
 	}
 
+	bool resizeLarge(shared(ExtentMap)* emap, Extent* e, size_t size) shared {
+		assert(e !is null, "Extent is null!");
+		assert(e.isLarge(), "Expected a large extent!");
+		assert(e.arenaIndex == index, "Invalid arena index!");
+		assert(isLargeSize(size));
+
+		auto computedPageCount = alignUp(size, PageSize) / PageSize;
+		uint newPages = computedPageCount & uint.max;
+
+		assert(newPages == computedPageCount, "Unexpected page count!");
+
+		uint oldSize = cast(uint) e.size;
+		uint oldPages = oldSize / PageSize;
+
+		// Huge is not supported:
+		if ((e.size > uint.max) || e.isHuge() || (oldPages > PageCount)) {
+			return false;
+		}
+
+		int pagesDelta = newPages - oldPages;
+		if (pagesDelta == 0) {
+			return true;
+		}
+
+		// Growing is not yet supported:
+		if (pagesDelta > 0) {
+			return false;
+		}
+
+		auto newSize = oldSize + pagesDelta * PageSize;
+		emap.clear(e.address + newSize, e.size - newSize);
+		shrinkAlloc(e, oldPages, -pagesDelta);
+		e.at(e.address, newSize, e.hpd);
```

Comment:
The size of the extent needs to be changed while the lock is held, no? In any case, if not (I'm not convinced) then the extent literally is not used in shrinkAlloc so why is it passed down?

lastly, it seems to me that the extent must be shrunk before the pages are cleared int he hpd, or we might see bad things happening.